TASK-79: Add friends backend

cristian-aldea opened the pull request:
closes #277

This PR aims to implement a few backend functions to be used in the features of #273.
   - addFriend
   - removeFriend
   - getRelationships

How to test:
- Make sure you have at least 2 users in your DB.
- Run the backend
- Try to use the addFriend and removeFriend mutations and check what happens when you call the `friends` query. It should behave as expected. addFriend should be a one-sided add, equivalent to sending a request. removeFriend should be a two-way remove.
- Call the getRelationships query. This returns a list of Relationships (see added class). What I recommend is going into MySQL and modifying the rows manually and making sure that the list returned by the query changes as expected.

Examples (all from perspective of user id 1):
No relationships:
![image](https://user-images.githubusercontent.com/23544999/112736721-3164fe00-8f2b-11eb-93ea-ccff72f44763.png)

user 1 added user 2
![image](https://user-images.githubusercontent.com/23544999/112736732-535e8080-8f2b-11eb-85de-16209abde242.png)

user 2 added user 1
![image](https://user-images.githubusercontent.com/23544999/112736737-640ef680-8f2b-11eb-8f70-a88ce6f4f7a7.png)

Both users added each other
![image](https://user-images.githubusercontent.com/23544999/112736752-7852f380-8f2b-11eb-8e26-6096a8753cd6.png)

Works with multiple users too
![image](https://user-images.githubusercontent.com/23544999/112736758-899c0000-8f2b-11eb-9310-6b1546306f3a.png)

cristian-aldea committed:
feat(friends): Implement friends backend functions

- Implemented removeFriend and getRelationships resolver + service methods
- Renamed UserRelationship to RelationshipType for clarity
- Added new output type Relationship
- Added utility methods for set operations

ref #277

cristian-aldea committed:
fix(graphql): Update graphql.tsx

ref #277

cristian-aldea committed:
fix(test): Fix and add tests for friends

- Reworked user mock data to not be dumb and repetitive

ref #277

cristian-aldea committed:
fix(e2e): Fix wrong start date during us-2.7 e2e

12:00 corresponds to 12pm in 12h format. 12am is expected. Fixed by replacing with 00:00

ref #277

r-saba commented on apps/mull-api/src/app/user/user.service.ts:
```diff
@@ -73,4 +75,87 @@ export class UserService {
   }
+
+  async removeFriend(currentUserId: number, userIdToRemove: number): Promise<boolean> {
+    if (currentUserId === userIdToRemove) {
+      throw new UnprocessableEntityException('Cannot remove oneself as friend.');
+    }
```
Should we also check that `userIdToRemove` is an actual user. Noticed we're not checking in the `addFriend` function as well which is a more serious issue as I believe it could break the UI.

cristian-aldea replied:
I think its not necessary, it would add an extra query (I think), and I don't see why we should send back an error if we want to remove someone from your friends who never was your friend to begin with (because they're a ghost)

r-saba replied:
Yes, I think it's overkill as well. Lets keep it as is.

r-saba commented on apps/mull-api/src/app/utilities.ts:
```diff
@@ -0,0 +1,11 @@
+export const setUnion = <T>(a: Set<T>, b: Set<T>) => {
```
Nice

cristian-aldea committed:
feat(user resolver): Set removeFriend to be two-way

I realized something odd in my design of removeFriend. removeFriend won't remove both sides of the relationship. This makes sense when the user wants to, say, cancel a friend request, but it makes less sense when the two users share a mutual friendship. Therefore I decided to have the method just remove the friendship on both sides in all cases

leojrsilao reviewed (approved):
LGTM!! 🚀 🚀 
Mutations and queries for adding friends are functional.
Unit tests and linters pass in the backend.

r-saba reviewed (approved):
LGTM! All queries, mutations,test and linter worked as expected.
